### PR TITLE
Add drive scope to BigQuery client

### PR DESCRIPTION
**Pull request by jeff303**
Adding new drive auth scope in order to be able to query Google Drive files created in BigQuery as external tables

Adding new test to confirm such a table can be queried successfully

Updating existing tests to work with new `project-id` and credentials

### CODE REVIEW

File: modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj [hunk 1]
```diff
@@ -33,17 +33,20 @@
...
+  `https://cloud.google.com/bigquery/external-data-drive`)."
+  ["https://www.googleapis.com/auth/bigquery"
+   "https://www.googleapis.com/auth/drive"])
 
 (defn- ^BigQuery database->client
   [database]
   (let [creds   (bigquery.common/database-details->service-account-credential (:details database))
         bq-bldr (doto (BigQueryOptions/newBuilder)
-                  (.setCredentials (.createScoped creds (Collections/singletonList bigquery-scope))))]
+                  (.setCredentials (.createScoped creds (LinkedList. bigquery-scopes))))]
```
Comment (camsaul) [inline modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:49]: Why not just make `bigquery-scopes` a list e.g. `(list ...)` and give it a `^clojure.lang.PersistentList` tag? Then you wouldn't need to convert it to a `LinkedList`. A Clojure `PersistentList` is a `java.util.List` after all

File: modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj [hunk 2]
```diff
@@ -33,17 +33,20 @@
...
+(def ^:private bigquery-scopes
+  "The scopes to use for executing BigQuery requests; see:
   `https://cloud.google.com/bigquery/docs/samples/bigquery-auth-drive-scope`.
-  Unclear if this can be sourced from the `com.google.cloud.bigquery` package directly."
-  "https://www.googleapis.com/auth/bigquery")
+  Unclear if this can be sourced from the `com.google.cloud.bigquery` package directly.  We use the standard bigquery
+  scope, as well as the drive scope (allowing for configured Drive external tables to be queried, as per
+  `https://cloud.google.com/bigquery/external-data-drive`)."
+  ["https://www.googleapis.com/auth/bigquery"
+   "https://www.googleapis.com/auth/drive"])
```
Comment (camsaul) [inline modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:43]:
```suggestion
  '("https://www.googleapis.com/auth/bigquery"
    "https://www.googleapis.com/auth/drive"))
```

File: modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj [hunk 3]
```diff
@@ -33,17 +33,20 @@
 ;;; |                                                     Client                                                     |
 ;;; +----------------------------------------------------------------------------------------------------------------+
 
-(def ^:private bigquery-scope
-  "The scope to use for executing BigQuery requests; see:
+(def ^:private bigquery-scopes
```
Comment (camsaul) [inline modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:36]:
```suggestion
(def ^:private ^clojure.lang.PersistentList bigquery-scopes
```

File: modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj (hunk 1 above)
Comment (camsaul) [reply modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:49]:
```suggestion
                  (.setCredentials (.createScoped creds bigquery-scopes)))]
```

**Review by camsaul — APPROVED** [review]
LGTM but I would just make `bigquery-scopes` a list in the first place so you don't need to convert it to a list every time you use it.

If you **did** convert it to a list `ArrayList` is probably a more appropriate choice than `LinkedList` anyway

File: modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj [hunk 4]
```diff
@@ -24,7 +24,7 @@
                                       BigQuery$TableOption BigQueryException BigQueryOptions Dataset DatasetId Field
                                       Field$Mode FieldValue FieldValueList QueryJobConfiguration Schema Table TableId
                                       TableResult]
-           java.util.Collections))
+           (clojure.lang PersistentList)))
```
Comment (camsaul) [inline modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:27]:
```suggestion
           clojure.lang.PersistentList))
```

Namespace checker is fussy and wants them done this way. Event tho it's the same thing

**Comment by jeff303** [comment]
Resolves #4179

**Comment by camsaul** [comment]
Sorry, you shouldn't have listened to me. `clojure.lang.PersistentList` has to come before all the BigQuery stuff.

FWIW for next time I personally wouldn't have bothered importing it and would have just used the full class name for the tag. I think it's a little more obvious in this case because you might be tempted to thing `PersistentList` is some sort of BigQuery thing

File: modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk_test.clj [hunk 5]
```diff
@@ -477,3 +477,22 @@
               (doseq [table (map Table [(u/the-id table1) (u/the-id table2)])]
                 ;; and the existing tables should have been updated with that schema
                 (is (= "my-dataset" (:schema table)))))))))))
+
+(deftest query-drive-external-tables
+  (mt/test-driver :bigquery-cloud-sdk
+    (testing "Google Sheets external tables can be queried via BigQuery"
```
Comment (camsaul) [inline modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk_test.clj]:
```suggestion
    (testing "Google Sheets external tables can be queried via BigQuery (#4179)"
```

**Comment by camsaul** [comment]
Hey @jeff303 make sure you double-check that the GH issues are actually closed out when your PR lands

**Comment by AyrtonB** [comment]
Thanks for adding this!

**Pull request merged**